internal/olm/operator/bundle/install.go

```diff
@@ -98,3 +168,169 @@ func (i *Install) setup(ctx context.Context) error {
+func addBundleToIndexImage(ctx context.Context, indexImage string, bundleDeclConfig *declarativeconfig.DeclarativeConfig) (*declarativeconfig.DeclarativeConfig, error) {
+	log.Infof("Rendering a File-Based Catalog of the Index Image %q", indexImage)
+	log.SetOutput(ioutil.Discard)
+	render := action.Render{
+		Refs: []string{indexImage},
+	}
+
+	imageDeclConfig, err := render.Run(ctx)
+	log.SetOutput(os.Stdout)
+	if err != nil {
+		return nil, fmt.Errorf("error rendering the index image %q: %v", indexImage, err)
+	}
+
+	var isPackagePresent, isBundlePresent, isChannelPresent bool
+	for _, pkg := range imageDeclConfig.Packages {
+		if pkg.Name == bundleDeclConfig.Packages[0].Name {
+			isPackagePresent = true
+			break
+		}
+	}
+
+	for _, bundle := range imageDeclConfig.Bundles {
+		if bundle.Name == bundleDeclConfig.Bundles[0].Name && bundle.Package == bundleDeclConfig.Bundles[0].Package {
+			isBundlePresent = true
+			break
+		}
+	}
+
+	for _, channel := range imageDeclConfig.Channels {
+		if channel.Name == bundleDeclConfig.Channels[0].Name && channel.Package == bundleDeclConfig.Bundles[0].Package {
+			isChannelPresent = true
+			break
+		}
+	}
+
+	// insert bundle to index if package/bundle/channel is not present on the index;
+	// and insert only bundle declarative config channel to index when both bundle/package are present but not the channel.
+	if !isPackagePresent && !isBundlePresent && !isChannelPresent {
+		imageDeclConfig.Bundles = append(imageDeclConfig.Bundles, bundleDeclConfig.Bundles[0])
+		imageDeclConfig.Channels = append(imageDeclConfig.Channels, bundleDeclConfig.Channels[0])
+		imageDeclConfig.Packages = append(imageDeclConfig.Packages, bundleDeclConfig.Packages[0])
+		imageDeclConfig.Others = append(imageDeclConfig.Others, bundleDeclConfig.Others...)
+	} else if isPackagePresent && isBundlePresent && !isChannelPresent {
+		// if the bundle already exists in a different channel,
+		// we already have the bundle and package blobs in the FBC, in which case,
+		// we need to reference it in the channel we want it to be in.
+		imageDeclConfig.Channels = append(imageDeclConfig.Channels, bundleDeclConfig.Channels[0])
+	} else if isBundlePresent {
+		return nil, fmt.Errorf("bundle %q already present in the index image: %s", bundleDeclConfig.Bundles[0].Name, indexImage)
+	} else {
+		return nil, fmt.Errorf("cannot add bundle %q to the index image: %s", bundleDeclConfig.Bundles[0].Name, indexImage)
+	}
+
+	log.Infof("Inserted the new bundle %q into the index image: %s", bundleDeclConfig.Bundles[0].Name, indexImage)
+
+	return imageDeclConfig, nil
+}
+
+// createFBC generates an FBC by creating bundle, package and channel blobs.
+func (f *FBCContext) createFBC(ctx context.Context) (bundleDeclcfg, error) {
+	var bundleDC bundleDeclcfg
+	// Rendering the bundle image into a declarative config format.
+	log.SetOutput(ioutil.Discard)
+	render := action.Render{
+		Refs: f.Refs,
```

Comment:
Do we expect `Refs` to always be length 1? If so, should we change the `FBCContext` struct to have a singular `Ref`?

And do we expect that ref to be a bundle? If so, let's set `action.Render`'s `AllowedRefMask`. You can check for [`ErrNotAllowed`](https://github.com/operator-framework/operator-registry/blob/aa8db09278a0ea30dd7cf07e3b197a3d7b5ce034/alpha/action/render.go#L49) when `Run()` returns and have your own error message along the lines of "only bundle images can be specified for `run bundle`"

Reply:
I'd prefer if we keep FBCContext as a string array. In some of the run bundle-upgrade work, we use this refs field for the previous injected bundles so that we can render all of them together.

Reply:
As @VenkatRamaraju mentioned, we are using the refs field to render both index image and bundle image, so the len varies for upgrade functionality.